### PR TITLE
Automatic flattening, fix for CopyLayer, comments

### PR DESCRIPTION
**Pull request by Atticus1806**
Since `CastLayer` is of instance `CopyLayer` it will be "skipped" during flattening which potentially causes dtype issues. This should fix the check, maybe its worth looking into the other subinstances of `CopyLayer` for other cases. 
From what I found these are: 
`DropoutLayer`, `ScaledGradientLayer`, `BatchNormLayer` and `TikhonovRegularizationLayer`. If you think only skipping `CopyLayer` would make sense I would change the whole check to:
`type(layer_) is CopyLayer`

Additionally I added some comments for some of the private functions of the flatten logic to make them more understandable, especially when seeing them first. 
Feel free to comment on them and make suggestions how to improve them, these are for now how I understood them and by far not perfect I think.
Also if you have a good suggestion for other functions in there we can add them too.

### CODE REVIEW

File: returnn/tf/network.py [hunk 1]
```diff
@@ -1480,7 +1486,6 @@ def _resolve_layer(layer_):
         continue
       blacklist.add(layer)
       layer_queue.extend(_layer_deps(layer))
-
```
Comment (albertz) [inline returnn/tf/network.py:1483]: Why did you remove this empty line?

File: returnn/tf/network.py [hunk 2]
```diff
@@ -1431,14 +1435,16 @@ def _check_push_flattening_to_inputs_for_layer(layer_):
 
     def _resolve_layer(layer_):
       """
+      Flattens the layer structure, removes irrelevant layers and returns next successor layer
       :param LayerBase layer_:
+      :return Next layer in succession
```
Comment (albertz) [inline returnn/tf/network.py:1440]:
```suggestion
      :return: next layer in succession
```

File: returnn/tf/network.py [hunk 3]
```diff
@@ -1337,6 +1337,7 @@ def _map_layer_dict_value(v):
 
     def _make_layer(layer_cls, layer_dict, map_opts=True):
       """
+      Creates the flattened layer
```
Comment (albertz) [inline returnn/tf/network.py:1340]: There should be an empty line after the description before the params (when the description is non-empty).

(Here and for all further docstrings as well.)

File: returnn/tf/network.py [hunk 4]
```diff
@@ -1431,14 +1435,16 @@ def _check_push_flattening_to_inputs_for_layer(layer_):
...
       :param LayerBase layer_:
+      :return Next layer in succession
       :rtype: LayerBase
       """
       while True:
         if isinstance(layer_, SubnetworkLayer):
           layer_ = layer_.subnetwork.layers["output"]
           continue
-        if isinstance(layer_, CopyLayer) and len(layer_.sources) == 1:
+        if isinstance(layer_, CopyLayer) and len(layer_.sources) == 1 and not isinstance(layer_, CastLayer):
```
Comment (albertz) [inline returnn/tf/network.py:1447]: I think instead of having such explicit check for `CastLayer`, we better should do just this:
```suggestion
        if type(layer_) is CopyLayer and len(layer_.sources) == 1:
```

File: returnn/tf/network.py [hunk 5]
```diff
@@ -1474,13 +1485,13 @@ def _resolve_layer(layer_):
       if layer in end_points:
         continue
       layer_queue.append(layer)
+
```
Comment (albertz) [inline returnn/tf/network.py:1488]: Why did you add this empty line?

File: returnn/tf/network.py (hunk 1 above)
Comment (albertz) [reply returnn/tf/network.py:1483]: Why was this removed?

File: returnn/tf/network.py (hunk 5 above)
Comment (Atticus1806) [reply returnn/tf/network.py]: "Restored" the wrong line, coffee not working yet :)

**Review by albertz — APPROVED** [review]
